crates/data_components/src/duckdb/write.rs

```diff
@@ -144,7 +156,15 @@ impl DataSink for DuckDBDataSink {
 
         let num_rows = match **self.duckdb.constraints() {
             [] => self.try_write_all_no_constraints(&tx, &data_batches)?,
-            _ => self.try_write_all_with_constraints(&tx, &data_batches)?,
+            _ => match self.try_write_all_with_constraints(&tx, &data_batches) {
+                Ok(num_rows) => num_rows,
+                Err(e) => {
+                    tx.commit()
```

Comment:
We need to commit the transaction here because `try_write_all_with_constraints` will create a temp DuckDB table to insert into, but if there is an error the delete table command is part of the transaction - so committing it will allow us to properly ensure we don't leave dangling temporary tables around.